Replace DeadlineTimer with boost::asio::deadline_timer (RIPD-1356)

scottschurr opened the pull request:
Replace uses of `DeadlineTimer` with `asio::steady_timer` in NetworkOPs and Application, (the only active users of `DeadlineTimer`).

This pull request is predicated on https://github.com/ripple/rippled/pull/2135, or something very much like it, going into develop.  The first three commits in this pull request are from https://github.com/ripple/rippled/pull/2135.  You don't need to review those for this pull request.  The (currently) two commits sitting atop them are the ones that need reviewing.

There are two things I'm particularly concerned about in this pull request:

- I'm an `asio` novice.  I may not be using `asio::steady_timer` correctly although my manual testing indicates it's doing the right thing.  Do I need an `asio::io_service::strand`?  Are there other details I need to be aware of?

- I made the former `JobCounter` more generic by allowing it to count closures with any set of argument and return types.  In the process I also made it support counting `lvalue` and `const lvalue` closures as well as `rvalue` closures.  I _think_ I got it right?  But my template work and move semantics should be carefully checked.  I don't live my life in templates.  Note that even though github's diff says ClosureCounter.h is a new file, it's really just a refinement with additional template parameters of the old JobCounter.h.

Reviewers: @miguelportilla for `asio` expertise, @HowardHinnant for template and move semantics expertise.

HowardHinnant commented on src/ripple/core/ClosureCounter.h:
```diff
@@ -0,0 +1,204 @@
+
+        Ret_t operator ()(Args_t... args)
+        {
+            return closure_ (args...);
```
I think this could be:
```
return closure_(std::forward<Args_t>(args)...);
```
Rationale:  If Args_t[i] is a non-reference type, this will move the args[i] from the by-value parameter list into the closure_ call, potentially avoiding an expensive copy.  If Args_t[i] is a lvalue-reference type, this will pass the args[i] by lvalue reference (const or non-const).  If Args_t[i] is an rvalue-reference type, this will move the args[i] into the closure, which is probably the intent of the client.

scottschurr commented:
Thanks @HowardHinnant!  I was so focused on getting the closure handling correct that I entirely spaced out the closure's arguments.  I've incorporated your fix and added a unit test to exercise argument handling.

scottschurr commented:
It turns out the previously added tests were not as thorough as I thought.  The most recent commit reworks the new tests.

miguelportilla commented on src/ripple/app/main/Application.cpp:
```diff
@@ -882,20 +882,44 @@ class ApplicationImp
-            m_jobQueue->addJob(
-                jtSWEEP, "sweep", [this] (Job&) { doSweep(); });
+            using namespace std::chrono_literals;
+            entropyTimer_.expires_from_now (5min);
```
I wonder if there is any benefit to making the interval configurable as the sweep.

miguelportilla replied:
I don't have any information about the rates either. I only mentioned it because the sweep timer is configurable and this one is not. A weak argument might be made that varying the entropy timer could increase difficulty for an attack knowing the server's time (interval is a given). I don't know if that is a case or not so I suggest leaving as is.

miguelportilla commented on src/ripple/app/main/Application.cpp:
```diff
@@ -882,20 +882,44 @@ class ApplicationImp
+        if (auto optionalCountedHandler = waitHandlerCounter_.wrap (
+            [this] (boost::system::error_code const& e)
+            {
+                if (e.value() == boost::system::errc::success)
```
Should we prevent setting the time if the application is stopping?

miguelportilla replied:
We may want to check if the application is stopping at this time before setting the timer.

miguelportilla replied:
In the event of a failure, we may want to log the error

miguelportilla commented on src/ripple/app/main/Application.cpp:
```diff
@@ -882,20 +882,44 @@ class ApplicationImp
+        if (auto optionalCountedHandler = waitHandlerCounter_.wrap (
+            [this] (boost::system::error_code const& e)
+            {
+                if ((e.value() == boost::system::errc::success) &&
```
In the event of a failure, we may want to log the error

miguelportilla commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -628,28 +630,40 @@ void NetworkOPsImp::setStateTimer ()
+    if (auto optionalCountedHandler = waitHandlerCounter_.wrap (
+        [this] (boost::system::error_code const& e)
+        {
+            if ((e.value() == boost::system::errc::success) &&
```
In the event of a failure, we may want to log the error and have some recourse to resetting the timer.

miguelportilla commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -628,28 +630,40 @@ void NetworkOPsImp::setStateTimer ()
+    if (auto optionalCountedHandler = waitHandlerCounter_.wrap (
+        [this] (boost::system::error_code const& e)
+        {
+            if ((e.value() == boost::system::errc::success) &&
```
In the event of a failure, we may want to log the error and have some recourse to resetting the timer.

miguelportilla commented on src/ripple/core/ClosureCounter.h:
```diff
@@ -0,0 +1,204 @@
+    std::mutex mutable mutex_ {};
+    std::condition_variable allClosuresDoneCond_ {};  // guard with mutex_
+    bool waitForClosures_ {false};                    // guard with mutex_
+    std::atomic<int> closureCount_ {0};
```
nit: can probably be unsigned.

miguelportilla commented on src/ripple/app/main/Application.cpp:
```diff
@@ -847,11 +847,11 @@ class ApplicationImp
-            m_sweepTimer.cancel ();
-            m_entropyTimer.cancel ();
-        }
+        sweepTimer_.cancel ();
```
It may be worth using the `cancel` overload with an error_code and log on failure.
eg.
```        boost::system::error_code ec;
        sweepTimer_.cancel(ec);
        if (ec)
        {
            JLOG (j_.trace()) << "Application: sweepTimer cancel error: " << ecCancel.message ();
        }```

miguelportilla commented on src/ripple/app/main/Application.cpp:
```diff
@@ -847,11 +847,11 @@ class ApplicationImp
-            m_entropyTimer.cancel ();
-        }
+        sweepTimer_.cancel ();
+        entropyTimer_.cancel ();
```
Same here

scottschurr commented:
This pull request is rebased, squashed, and now incorporates the passed version of https://github.com/ripple/rippled/pull/2135.

scottschurr commented:
I'm calling this pull request successfully rebased, even though Travis isn't building.  If I drop these commits on top of https://github.com/ripple/rippled/pull/2190 then that conglomeration passes all continuous integration checks including Travis.

seelabs commented:
In 0.80.0-b4
